### ISSUE TITLE
Extract vertical levels in preprocessing of model data

### ISSUE THREAD

**Issue by tsjackson-noaa** [NOAA-GFDL/MDTF-diagnostics#40]
Many PODs require (and assume) model data on atmospheric pressure levels, but CMIP6 and other data sources provide four-dimensional data. Level extraction is currently handled on an ad-hoc, per-POD basis. Handling level extraction in the framework is more in keeping with its design goals.

There are several aspects to this task:

- Remove current use of [NCO](http://nco.sourceforge.net/) to preprocess model data files (determine coordinate names, truncate data to analysis period, concatenate multi-file chunks, etc.) in favor of [xarray](http://xarray.pydata.org/en/stable/).
- Parse and handle extended variable description syntax in the POD [settings file](https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/ref_settings.html).
- Implement appropriate vertical interpolation for [parametric vertical coordinates](http://cfconventions.org/cf-conventions/cf-conventions.html#parametric-v-coord), in particular the hybrid sigma coordinate used in NCAR CAM/CESM.
- Merge the separate [feature/gfdl-data](https://github.com/NOAA-GFDL/MDTF-diagnostics/tree/feature/gfdl-data) branch into [main](https://github.com/NOAA-GFDL/MDTF-diagnostics). The former branch contains per-POD modifications for 4D data needed to test the framework on CMIP6 data hosted at GFDL. This will be opened as a separate issue.

**Comment by jkrasting** [comment]
@tsjackson-noaa - thanks. This will be a great feature to have in the framework.  Two questions:

1. how exactly would a POD signal to the framework in the settings file that it would like extracted data?
2. could PODs still accept 4D data if they had their own routines? (i.e. vertical integration for example)

**Comment by tsjackson-noaa** [comment]
1. The syntax I proposed for level extraction is the `scalar_coordinates` attribute for variable description: https://mdtf-diagnostics.readthedocs.io/en/latest/sphinx/ref_settings.html#varlist-entry-example . We may want to revisit this if it's insufficiently general.

2. For every preprocessing ability we add to the framework, we should add an optional flag to the settings file allowing PODs to disable that preprocessing. At the same time, we want to encourage new PODs to be as specific as possible in the data they require.

**Comment by tsjackson-noaa** [comment]
Cross-referencing the Eulerian Storm Tracker POD (#52), which also needs level extraction.

**Comment by jkrasting** [comment]
@tsjackson-noaa - just checked out `feature/level-extraction` from your fork.

I created some synthetic GFDL CMIP output to use with the Wheeler-Kiladis diagnostic.  It generates `pr`,`rlut` as 2D fields and `wap`, `ua`, and `va` as 3D variables on the plev19 vertical coordinate.  This is now available on the main branch of [jkrasting/mdtf_test_data](https://github.com/jkrasting/mdtf_test_data)

I tried running your branch with this group of settings:
```
      "CASENAME" : "GFDL.Synthetic",
      "model" : "GFDL",
      "convention" : "CMIP",
      "FIRSTYR" : 1977,
      "LASTYR" : 1981,
      "pod_list": [
          // Optional: PODs to run for this model only (defaults to all)
          "Wheeler_Kiladis"
      ]   
    }  
```

And I ran into this error:

```
======= Starting /Users/krasting/mdtf/MDTF-diagnostics/src/mdtf.py
####### alts for {new_v.name}:
omega500_var alternate
####### alts for {new_v.name}:
u200_var alternate
####### alts for {new_v.name}:
u850_var alternate
	Querying 'U850 @ 1day'
	Caught exception querying U850 @ 1day: DataQueryError(DataKey(case_name='GFDL.Synthetic', date_range=DateRange('1977-1981'), name_in_model='U850', frequency=DateFrequency('1day'), level=None), 'File not found at /Users/krasting/mdtf/inputdata/model/GFDL.Synthetic/day/GFDL.Synthetic.U850.day.nc.').
	Querying 'pr @ 1day'
	Querying 'rlut @ 1day'
	Querying 'OMEGA500 @ 1day'
	Caught exception querying OMEGA500 @ 1day: DataQueryError(DataKey(case_name='GFDL.Synthetic', date_range=DateRange('1977-1981'), name_in_model='OMEGA500', frequency=DateFrequency('1day'), level=None), 'File not found at /Users/krasting/mdtf/inputdata/model/GFDL.Synthetic/day/GFDL.Synthetic.OMEGA500.day.nc.').
	Querying 'U200 @ 1day'
	Caught exception querying U200 @ 1day: DataQueryError(DataKey(case_name='GFDL.Synthetic', date_range=DateRange('1977-1981'), name_in_model='U200', frequency=DateFrequency('1day'), level=None), 'File not found at /Users/krasting/mdtf/inputdata/model/GFDL.Synthetic/day/GFDL.Synthetic.U200.day.nc.').
	Wheeler_Kiladis: request for 'omega500_var (= OMEGA500) @ 1day, 500 hPa' failed; finding alternate vars.
	Wheeler_Kiladis: no alternates available for 'omega500_var (= OMEGA500) @ 1day, 500 hPa'.
	Wheeler_Kiladis: request for 'u200_var (= U200) @ 1day, 200 hPa' failed; finding alternate vars.
	Wheeler_Kiladis: no alternates available for 'u200_var (= U200) @ 1day, 200 hPa'.
	Wheeler_Kiladis: request for 'u850_var (= U850) @ 1day, 850 hPa' failed; finding alternate vars.
	Wheeler_Kiladis: no alternates available for 'u850_var (= U850) @ 1day, 850 hPa'.
Creating /Users/krasting/mdtf/wkdir/MDTF_GFDL.v6.Synthetic_1977_1981.tar.
	DEBUG: cleanup temp dir /Users/krasting/mdtf/wkdir/MDTF_temp_1f22294f0642fce1
	DEBUG: cleanup temp dir /Users/krasting/mdtf/wkdir/MDTF_temp_6c9dbc282f75dd7a
	DEBUG: cleanup temp dir /Users/krasting/mdtf/wkdir/MDTF_temp_x62bffaeec9601d03

Exiting with errors from /Users/krasting/mdtf/MDTF-diagnostics/src/mdtf.py
The following PODs raised errors: Wheeler_Kiladis
Output written to /Users/krasting/mdtf/wkdir
```

**Comment by tsjackson-noaa** [comment]
Closing this issue since it has been implemented as of the [v3.0 beta 3](https://github.com/NOAA-GFDL/MDTF-diagnostics/releases/tag/v3.0-beta.3) release. See PR NOAA-GFDL/MDTF-diagnostics#147.
